### PR TITLE
Replacement-based JS instrumentation (close #135)

### LINKED ISSUE #135: Replacement-based JS instrumentation
What if we'll modify our code instrumentation, so it will not regenerate whole AST, but just chunks what we need to replace:

1. Parse JS with the node location info
2. Find nodes which we need to modify
3. Modify node and instantly generate JS code for it, when using location info replace substrings in the source code.

Pros:
- Performance. Code generation is a bottle neck in the whole mechanism, reducing it to just individual nodes should give us significant performance gain
- Quality. We will preserve original formatting and keep comments in place (which is sometimes quite important, e.g.: bugs with `func.toString()`)

\cc @churkin

### PR DESCRIPTION
**Pull request by churkin**
\cc @inikulin @miherlosev

### CODE REVIEW

**Comment by churkin** [comment]
@testcafe-build-bot \retest

File: src/processing/script/transformers/empty-function.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+// Transform:
+// function() {} --> function() {__emptyBody$Marker;});
+
+export default {
+    nodeReplacementRequireTransform: false,
+
+    nodeTypes: [Syntax.BlockStatement],
+
+    condition: (node, parent) => {
```
Comment (inikulin) [inline src/processing/script/transformers/empty-function.js]: BTW, why we need this transform?

File: src/processing/script/index.js [hunk 2]
```diff
@@ -117,19 +112,45 @@ export function isScriptProcessed (code) {
...
+
+    chunks.push(script.substring(index));
+
+    return chunks.join('');
+}
+
+export function processScript (src, withHeader) {
     var { bom, preprocessed } = preprocess(src);
-    var { ast, isObject }     = analyze(preprocessed);
+    var withoutHtmlComments = removeHtmlComments(preprocessed);
```
Comment (inikulin) [inline src/processing/script/index.js]: Formatting

File: src/processing/script/transformers/node-replacer.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
+// -------------------------------------------------------------
+// WARNING: this file is used by both the client and the server.
+// Do not use any browser or node-specific API!
+// -------------------------------------------------------------
+
+export default function replace (node, newNode, parent, key) {
```
Comment (inikulin) [inline src/processing/script/transformers/node-replacer.js]: It should be called `replaceNode` and file should be called `replace-node.js`

**Comment by inikulin** [comment]
\r-

File: src/processing/script/index.js (hunk 2 above)
Comment (churkin) [inline src/processing/script/index.js]: It seems that the WebStorm is not able to properly format similar constructions :(

**Comment by churkin** [comment]
FPR

**Comment by inikulin** [comment]
lgtm

**Comment by miherlosev** [comment]
lgtm.
Waiting for Health Monitor results...

**Pull request merged**